corehq/apps/userreports/specs.py

```diff
@@ -18,15 +19,70 @@ def TypeProperty(value):
     return StringProperty(required=True, choices=[value])
 
 
-@dataclass
-class FactoryContext():
-    named_expressions: dict
-    _named_expressions: dict = field(init=False, repr=False)
+class BaseContainer:
+    """Class used to manage access to named filters and expressions and
+    check for circular references.
+    """
+    def __init__(self, factory_context, expressions, domain=None):
+        self.factory_context = weakref.proxy(factory_context)
+        self.expressions = expressions
+        self.domain = domain
+        self.stack = []
+
+    def replace(self, expressions):
+        self.expressions = expressions
+
+    def get(self, name):
+        with self._search_scope(name):
+            from corehq.apps.userreports.models import LazyExpressionWrapper
+            all_expressions = self._get_db_expressions() | self.expressions
+            expr = all_expressions.get(name)
+            if not expr:
+                raise BadSpecError(gettext("Couldn't find named expression with name: {name}").format(name=name))
```

Comment:
This error message always uses the word "expression", but sometimes it's a _filter_ that's being looked up, right?

Another nudge that this `get` method should be renamed to `__getitem__` and it should raise `KeyError` here.

Reply:
I debated about correcting that but it didn't seem necessary as the names provide clear indication of what was missing. Raising a `KeyError` would require handling the error by the callers which just convert it to another error anyway. But, I can push the error handling up one level to the `FactoryContext` which solves both issues: de4b0b5f35f29ee579b66a3179fef2d87b909d47